### ISSUE TITLE
Current position moves out of the map during navigation 

### ISSUE THREAD

**Issue by rinigus** [otsaloma/poor-maps#25]
I had a chance to test the navigation feature of Poor Maps today. The version that I used, contains 24dd599620b8a9cf80afad49ece7c40b579525c2 (Account for navigation block and auto-rotate when centering) already. 

The use was as follows: OSM Scout as a router, select Navigate to (address specified) from current position. The interaction on setup was done in portrait mode, started navigation, phone turned to landscape, and started driving. Starting from some moment, the current position indicator was getting either very close to the upper message or, after some time, disappeared. Now, I am not 100% sure whether I touched the screen or something touched it while driving. After disappearance, I couldn't get the location indicator back (well, I didn't try very hard while driving). 

Maybe I did something wrong or should have setup it somehow differently. Please let me know.

If you don't mind suggestion - would be great to get the current location fixed at the middle of the screen horizontally and on 1/4'th of the screen vertically (maybe should play with the position a bit or allow user to set it). That way the map would move with the movement keeping the current location in the same spot. 

Also, I don't know if you have any algorithm to calculate the shown zoom level on the map, or is that fixed by user when starting navigation?

Please let me know if opening an issue is a correct way of sending such feedback or whether you would prefer TMO.

Thank you for a great application!

**Comment by otsaloma** [comment]
It sounds like you didn't have auto-center on at all. There's three ways to turn on auto-center: (1) tap on the position icon, (2) at the bottom of the main menu or (3) the best option when navigating, tap on the narrative block at the top of the screen and then tap "Begin". That will turn on auto-center, auto-rotate and set a sensible zoom level. This is based on the assumed interaction flow that the user first examines the whole route to see that it's OK and then hits some button when ready to go.

The discoverability of that preferred option (3) is probably bad -- maybe that narrative block should display "Tap to begin" or something initially?

> Also, I don't know if you have any algorithm to calculate the shown zoom level on the map, or is that fixed by user when starting navigation?

The Begin button sets the zoom level at 16, if currently less than that. The user can then change that to anything he wants. It doesn't change at all during navigation. I'm aware that it should perhaps zoom into intersections and zoom out between, but I haven't gotten around to thinking about the details of how that should work.

> Please let me know if opening an issue is a correct way of sending such feedback or whether you would prefer TMO.

TMO is better for wider announcements and gathering opinions, but for asking me specifically, these issues are better. Or, if you have lots of small miscellaneous questions I can also set up a Gitter room, I have those for a couple other projects.

**Comment by rinigus** [comment]
I am actually pretty sure I used option (3) [I didn't realize that there are options (1) and (2)]. So, when I started driving, I pressed Begin. But let me test it again and see if I can reproduce it. It would probably have to wait till Monday, though.

In general, option (3) seems to be rather simple to find and I did examine the route. But I think this 'Tap to begin' I have seen in some programs and its probably to ensure that the user would do that.

Re zoom: its probably OK as it is. The zooming in / out could end up rather confusing.

To sum up, I'll test again when I get a chance and would report back on it.

**Comment by otsaloma** [comment]
OK. It's also possible I broke something in 24dd599620b8a9cf80afad49ece7c40b579525c2. I haven't field-tested that yet, I expect to do so on wednesday or thursday when I should have the new auto-rotate mechanism as well.

**Comment by otsaloma** [comment]
It seems I did break the auto-centering. I forgot to account for the rotation angle, so the new code only works if the map is not rotated. I'll fix it.

**Comment by otsaloma** [comment]
If I got the trigonometry right, it should work now: b027784efb5d0f45c877ba6039225309c48fa020

**Comment by rinigus** [comment]
I managed to test it today on a shorter trip and all was working as expected. So, it looks that this commit resolved the issue. Thank you!

**Comment by otsaloma** [comment]
Good to hear, thanks for testing.

**Comment by rinigus** [comment]
Thank you! I'll test it tomorrow.

**Comment by rinigus** [comment]
Thank you for letting me know! Would you mind to add also an option that
would allow to move the current position somewhat lower from the center?
While you drive, most important is usually towards the front of the
movement...

Best wishes,

rinigus

**Comment by otsaloma** [comment]
I don't add prefences for that kind stuff, [it's not the right approach](http://ometer.com/free-software-ui.html). I try to get the defaults right instead of leaving it to each user to fix. Currently if auto-rotate is on, the centering is shifted 10 % down on screen. That's a preliminary figure -- I haven't tested this properly yet, I'll do so tomorrow. It's well possible it could be more, depends mostly on how well the new auto-rotate mechanism works.

**Comment by rinigus** [comment]
Well, knowing the result of your approach and admiring it - I don't mind if you test and find the best solution. It could be that the "ideal position" is somewhat different for car and walking.

**Comment by otsaloma** [comment]
The new auto-rotate worked flawless in my testing, so there's no obstacle to lowering the centering point. Done in bcc649cd5ede0a8e87627de8bfebf82c80627d12.

Here's what the center-points and the radius, crossing which new auto-centering is triggered, looks like with and without auto-rotate:

![screenshot-20161026-233832](https://cloud.githubusercontent.com/assets/237265/19744381/ff4897b2-9bd5-11e6-94ba-c3376bfce987.png)

![screenshot-20161026-233844](https://cloud.githubusercontent.com/assets/237265/19744382/ff6d1510-9bd5-11e6-89a2-677c6f807f06.png)

I'll still need to test this. If it looks like it's centering too often, I might bump the radius a bit as the constant recentering could waste battery.

**Comment by rinigus** [comment]
Looks great! Thank you.
